eng-wasm/derive/src/pub_interface.rs

```diff
@@ -0,0 +1,475 @@
+    let function_name_func_name = FUNCTION_NAME_FUNC_NAME.into_ident();
+    let args_func_name = ARGS_FUNC_NAME.into_ident();
+    let call_func_name = CALL_FUNC_NAME.into_ident();
+
+    let aux_functions = generate_eng_wasm_aux_functions(&function_name_func_name, &args_func_name);
+    let constructor_function =
+        generate_deploy_function(&pub_interface_signatures, &deploy_func_name);
+    let dispatch_function =
+        generate_dispatch_function(&dispatch_func_name, &pub_interface_signatures);
+
+    quote! {
+        #item
+        #aux_functions
+        #constructor_function
+        #dispatch_function
+
+        #[no_mangle]
+        pub fn #call_func_name(){
+            #dispatch_func_name(&#function_name_func_name(), &#args_func_name());
+        }
+    }
+}
+
+fn generate_eng_wasm_aux_functions(
+    function_name_func_name: &syn::Ident,
+    args_func_name: &syn::Ident,
+) -> proc_macro2::TokenStream {
+    quote! {
+        #[no_mangle]
+        pub fn #function_name_func_name() -> eng_wasm::String {
+            use eng_wasm::{
+                Vec,
+                String,
+                ToString,
+                from_utf8,
+                external::{
+                    fetch_function_name,
+                    fetch_function_name_length,
+                },
+            };
+
+            let length_result = unsafe { fetch_function_name_length() };
+
+            match length_result {
+                0 => String::new(),
+                length => {
+                    let mut data = Vec::with_capacity(length as usize);
+                    for _ in 0..length{
+                        data.push(0);
+                    }
+
+                    unsafe {
+                        fetch_function_name(data.as_mut_ptr());
+                    }
+                    from_utf8(&data).unwrap().to_string()
+                }
+            }
+        }
+
+        #[no_mangle]
+        pub fn #args_func_name() -> eng_wasm::Vec<u8> {
+            use eng_wasm::{
+                Vec,
+                external::{
+                    fetch_args,
+                    fetch_args_length,
+                },
+            };
+            let length_result = unsafe { fetch_args_length() };
+
+            match length_result {
+                0 => Vec::new(),
+                length => {
+                    let mut data = Vec::with_capacity(length as usize);
+                    for _ in 0..length{
+                        data.push(0);
+                    }
+
+                    unsafe {
+                        fetch_args(data.as_mut_ptr());
+                    }
+                    data
+                }
+            }
+        }
+    }
+}
+
+fn generate_deploy_function(
+    signatures: &PubInterfaceSignatures,
+    deploy_func_name: &syn::Ident,
+) -> proc_macro2::TokenStream {
+    if let Some(constructor_signature) = signatures
+        .signatures
+        .iter()
+        .find(|item| item.ident == CONSTRUCTOR_NAME)
```

Comment:
Please add one more check about the constructor: it should return nothing.

Reply:
I added this check, as well as a check that the constructor is marked as public when defined in an `impl`